api/helpers/oauth/fetch-discord-identity.js

```diff
@@ -0,0 +1,66 @@
+module.exports = {
+  friendlyName: 'Generate Secret',
+
+  description: 'Generates a secret for OAuth state',
+
+  inputs:{
+    token: {
+      type: 'ref',
+      description: 'Discord Access Token',
+      required: true
+    },
+    user: {
+      type: 'ref',
+      description: 'Exisiting User'
+    }
+  },
+
+  fn: async function ({ token, user }, exits) {
+    try {
+      const response = await fetch('https://discord.com/api/users/@me', {
+        headers: {
+          'Authorization': `Bearer ${token.access_token}`,
+        },
+      });
+
+      if (!response.ok) {
+        throw new Error(`Failed to fetch user info: ${response.status}`);
+      }
+
+      const data = await response.json();
+
+      const providerId = data.id;
+      const provider = 'discord';
+      const { email } = data;
+      let { username } = data;
+
+      const prevIdentity = await Identity.findOne({ providerId }).populate('user');
+
+      if (prevIdentity && prevIdentity.user) {
+        if(user && prevIdentity.user.id !== user){
+          return exits.error('Account already linked to another profile');
+        }
+        return exits.success(prevIdentity.user);
+      }
+      const existingUserName = await User.findOne({ username });
+
+      if (existingUserName) {
+        username = `${username}${Math.floor(Math.random() * 1000)}`;
+      }
+      const updatedUser = user ? await User.findOne({ id: user }) : await User.create({ username }).fetch();
```

Comment:
So if their username from the oauth provider is already available, we automatically assign it to them, and if that username already exists in our database, we append a random number to it?

It's intriguing, but let's bake the UX of it. I think we're going to want to give users the ability to change their username from their discord one. The question is whether we should just make changing your username a thing, generally, or specifically make it a part of this flow